python-lib/config.py

```diff
@@ -82,27 +82,38 @@ def get_hdfs_location(dataset_config: Mapping[AnyStr, Any], sf_stage_name: AnySt
     return f'{sf_stage_name}{path}/'
 
 
-def get_snowflake_to_hdfs_query(sf_location: AnyStr, sf_table_name: AnyStr) -> AnyStr:
+def get_snowflake_to_hdfs_query(sf_location: AnyStr, sf_table_name: AnyStr,
+                                sf_schema: List[Mapping[AnyStr, AnyStr]]) -> AnyStr:
     """
     Gets a COPY statement for copying from a Snowflake Table to an HDFS location
-    :param sf_location:
-    :param sf_table_name:
+    :param sf_location: HDFS location
+    :param sf_table_name: Snowflake table
+    :param sf_schema: Snowflake schema
     :return: a SQL COPY statement
     """
     # moving this function here isn't strictly necessary as it's not re-used, but it makes
     # things significantly easier to unit test.
 
+    # Generate SELECT clause and cast TIMESTAMP_TZ, TIMESTAMP_LTZ to TIMESTAMP
+    # This is required as COPY command does not support TZ and LTZ. See AOBS-491 for more details
+    columns = ''
+    for col in sf_schema:
+        columns += '"'+col['name']+'"'
+        if col['originalType'] in ['TIMESTAMPLTZ', 'TIMESTAMPTZ']:
+            columns += '::TIMESTAMP AS "'+col['name']+'"'
+        columns += ','
```

Comment:
Once you've moved the column schema mapping out of this function, you can more easily use list comprehension and `', '.join()` to create this column selection. Preferably in the same spot that the other query generator does it.

https://github.com/outrauk/dataiku-plugin-snowflake-hdfs/blob/9a34965b4ff41951bcf163ef9919cf2adc14089a/python-lib/config.py#L119-L127

Reply:
Something like this:

```python
columns = [f'"{c["name"]}"::TIMESTAMP AS "{c["name"]}"' 
           if c["originalType"] in ['TIMESTAMPLTZ', 'TIMESTAMPTZ'] 
           else f'"{c["name"]}"' for c in sf_schema
          ]

    sql = f"""
COPY INTO '{sf_location}'
FROM (
    SELECT {', '.join(columns)}
    FROM {sf_table_name}
)
FILE_FORMAT = (TYPE = PARQUET)
OVERWRITE = TRUE
HEADER = TRUE;
    """
```

Reply:
Or if list comprehension isn't your style, something like this:

```python
def wrap_column(c):
    if c["originalType"] in ['TIMESTAMPLTZ', 'TIMESTAMPTZ']:
        return f'"{c["name"]}"::TIMESTAMP AS "{c["name"]}"'
    else:
        return f'"{c["name"]}"'

columns = map(wrap_column, sf_schema)

sql = f"""
COPY INTO '{sf_location}'
FROM (
    SELECT {', '.join(columns)}
    FROM {sf_table_name}
)
FILE_FORMAT = (TYPE = PARQUET)
OVERWRITE = TRUE
HEADER = TRUE;
    """
```

Reply:
Yes - list comprehension is best suited here imho.